lib/fuse_uring.c

```diff
@@ -0,0 +1,571 @@
+
+	struct io_uring ring;
+
+	/* size depends on queue depth */
+	struct fuse_ring_ent ent[];
+};
+
+/**
+ * Main fuse_ring structure, holds all fuse-ring data
+ */
+struct fuse_ring_pool {
+	struct fuse_session *se;
+
+	size_t nr_queues;  /* number of queues */
+	size_t queue_depth; /* number of per queue entries */
+	size_t max_req_payload_sz;
+	size_t queue_mem_size;
+	struct fuse_ring_queue *queues;
+};
+
+static size_t
+fuse_ring_queue_size(const size_t q_depth)
+{
+	const size_t req_size = sizeof(struct fuse_ring_ent) * q_depth;
+
+	return sizeof(struct fuse_ring_queue) + req_size;
+}
+
+static struct fuse_ring_queue *
+fuse_uring_get_queue(struct fuse_ring_pool *fuse_ring, int qid)
+{
+	char *ptr =
+		((char *)fuse_ring->queues) + (qid * fuse_ring->queue_mem_size);
+
+	return (struct fuse_ring_queue *)ptr;
+}
+
+/**
+ * return a pointer to the 80B area
+ */
+static void *fuse_uring_get_sqe_cmd(struct io_uring_sqe *sqe)
+{
+	return (void *)&sqe->cmd[0];
+}
+
+static void fuse_uring_sqe_set_req_data(struct fuse_uring_cmd_req *req,
+					const unsigned int qid,
+					const unsigned int commit_id)
+{
+	req->qid = qid;
+	req->commit_id = commit_id;
+	req->flags = 0;
+}
+
+static void
+fuse_uring_sqe_prepare(struct io_uring_sqe *sqe, struct fuse_ring_ent *req,
+		       __u32 cmd_op)
+{
+	/* These fields should be written once, never change */
+	sqe->opcode = IORING_OP_URING_CMD;
+
+	/*
+	 * IOSQE_FIXED_FILE: fd is the index to the fd *array*
+	 * given to io_uring_register_files()
+	 */
+	sqe->flags = IOSQE_FIXED_FILE;
+	sqe->fd = 0;
+
+	sqe->rw_flags = 0;
+	sqe->ioprio = 0;
+	sqe->off = 0;
+
+	io_uring_sqe_set_data(sqe, req);
+
+	sqe->cmd_op = cmd_op;
+	sqe->__pad1 = 0;
+}
+
+static int fuse_queue_setup_io_uring(struct io_uring *ring, size_t qid,
+				     size_t depth, int fd, int evfd)
+{
+	int rc;
+	struct io_uring_params params = {0};
+	int files[2] = { fd, evfd };
+
+	depth += 1; /* for the eventfd poll SQE */
+
+	params.flags = IORING_SETUP_SQE128;
+
+	/* Avoid cq overflow */
+	params.flags |= IORING_SETUP_CQSIZE;
+	params.cq_entries = depth * 2;
+
+	/* These flags should help to increase performance, but actually
+	 * make it a bit slower - reason should get investigated.
+	 */
+	if (0) {
+		/* Has the main slow down effect */
+		params.flags |= IORING_SETUP_SINGLE_ISSUER;
+
+		// params.flags |= IORING_SETUP_DEFER_TASKRUN;
+		params.flags |= IORING_SETUP_TASKRUN_FLAG;
+
+		/* Second main effect to make it slower */
+		params.flags |= IORING_SETUP_COOP_TASKRUN;
+	}
+
+	rc = io_uring_queue_init_params(depth, ring, &params);
+	if (rc != 0) {
+		fuse_log(FUSE_LOG_ERR, "Failed to setup qid %zu: %d (%s)\n",
+			 qid, rc, strerror(-rc));
+		return rc;
+	}
+
+	rc = io_uring_register_files(ring, files, 1);
+	if (rc != 0) {
+		rc = -errno;
+		fuse_log(FUSE_LOG_ERR,
+			 "Failed to register files for ring idx %zu: %s",
+			 qid, strerror(errno));
+		return rc;
+	}
+
+	return 0;
+}
+
+static void free_payload(struct fuse_ring_ent *ent)
```

Comment:
free_payload() has only one caller, thus better to embed it in the caller for readability.

Reply:
Thanks, left over from my other development branch that has support for external threads and payload allocation through these threads (needed to get the RDMA key). The API for that will be solved differently - I want to have the application to own the ring and libfuse to able to submit requests to it.